### PR TITLE
feat: CLI support for switching between multiple logged in user accounts

### PR DESCRIPTION
**Pull request by quinton11**
See #340. 

This feature allows users  login with multiple accounts and switch between them using [infisical switch] without having to re-authenticate anytime they intend to switch scopes.

### CODE REVIEW

**Comment by maidul98** [comment]
Thanks for the PR. I will review it this week!

**Comment by maidul98** [comment]
Just reviewd the PR, works great! I'm going to make some minor changes and then i'll merge it

**Comment by maidul98** [comment]
Just wanted to also mention that the community has been asking for the host API to be persisted, and I think it's a good idea. By attaching the domain to use for APIs when we add a new user, we can fulfill this request. It would be great if we could do this sooner rather than later because we'll need to make sure everything is compatible with the Config data structure we save. Would you like to give it a shot? If you're busy, no worries, just let me know, and I'm happy chime in on it.

I think what should happen is we should allow the user to choose a host, the default being infisical cloud before login and then continue the login steps. Once done, save the host along with the user email and then we can change this to point to the config https://github.com/Infisical/infisical/blob/main/cli/packages/cmd/root.go#L35

**Comment by quinton11** [comment]
Great, it wouldn't be a problem.. I'll look into it. Thanks!

> Just wanted to also mention that the community has been asking for the host API to be persisted, and I think it's a good idea. By attaching the domain to use for APIs when we add a new user, we can fulfill this request. It would be great if we could do this sooner rather than later because we'll need to make sure everything is compatible with the Config data structure we save. Would you like to give it a shot? If you're busy, no worries, just let me know, and I'm happy chime in on it.
> 
> I think what should happen is we should allow the user to choose a host, the default being infisical cloud before login and then continue the login steps. Once done, save the host along with the user email and then we can change this to point to the config https://github.com/Infisical/infisical/blob/main/cli/packages/cmd/root.go#L35

**Comment by quinton11** [comment]
Changes:
On user logins, user's are prompted for the preffered hosting domain associated with the account (default being infisical cloud). Config data structure supports multiple users along with their domains. 
An added user sub command allowing users change the domain for a logged in account (in case of a domain url change)

File: cli/packages/cmd/login.go [hunk 1]
```diff
@@ -48,8 +54,17 @@ var loginCmd = &cobra.Command{
 			util.HandleError(err)
 		}
 
+		// addUser := false
```
Comment (maidul98) [inline cli/packages/cmd/login.go:57]: I think you forgot to delete these?

File: cli/packages/cmd/login.go [hunk 2]
```diff
@@ -252,6 +274,51 @@ func init() {
...
+func askForDomain() error {
+	//query user to choose between Infisical cloud or self hosting
+	options := []string{"Infisical Cloud", "Self Hosting"}
+	optionsPrompt := promptui.Select{
+		Label: "Select your hosting option",
+		Items: options,
+		Size:  2,
+	}
+
+	idx, _, err := optionsPrompt.Run()
```
Comment (maidul98) [inline cli/packages/cmd/login.go:286]: nit: is there a reason why you have choosen `idx`? I would think it's better to call it something like `selectedHostingOption`

File: cli/packages/cmd/login.go [hunk 3]
```diff
@@ -252,6 +274,51 @@ func init() {
...
+		Validate: urlValidation,
+	}
+
+	domain, err := domainPrompt.Run()
+	if err != nil {
+		return err
+	}
+
+	//set api url
+	config.INFISICAL_URL = domain
```
Comment (maidul98) [inline cli/packages/cmd/login.go:317]: we should add a `/api` at the end because all of our default nginx configs route traffic from `/api` to the backend services

File: cli/packages/cmd/login.go [hunk 4]
```diff
@@ -252,6 +274,51 @@ func init() {
...
+		_, err := url.ParseRequestURI(input)
+		if err != nil {
+			return errors.New("this is an invalid url")
+		}
+		return nil
+	}
+
+	//else run prompt to enter domain
+	domainPrompt := promptui.Prompt{
+		Label:    "Domain",
```
Comment (maidul98) [inline cli/packages/cmd/login.go:307]: I wonder if domain is clear, when i think of domain i think of google.com not the https or http before it. Do you know if there is a way to provide a placeholder text with an example so users know?

File: cli/packages/cmd/login.go [hunk 5]
```diff
@@ -342,16 +409,16 @@ func getFreshUserCredentials(email string, password string) (*api.GetLoginOneV2R
 	return &loginOneResponseResult, &loginTwoResponseResult, nil
 }
 
-func shouldOverrideLoginPrompt(currentLoggedInUserEmail string) (bool, error) {
+func userLoginMenu(currentLoggedInUserEmail string) (bool, error) {
 	prompt := promptui.Select{
-		Label: fmt.Sprintf("There seems to be a user already logged in with the email: %s. Would you like to override that login? Select[Yes/No]", currentLoggedInUserEmail),
-		Items: []string{"No", "Yes"},
+		Label: fmt.Sprintf("Current logged in user email: %s", currentLoggedInUserEmail),
```
Comment (maidul98) [inline cli/packages/cmd/login.go:414]: what do you think about also adding which host they are on such as `current logged in user email: maidul@ss.com on domain: app.infisical.com`?

File: cli/packages/cmd/user.go [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+	},
+}
+
+func init() {
+	userCmd.AddCommand(domainCmd)
+	userCmd.AddCommand(switchCmd)
+	rootCmd.AddCommand(userCmd)
+}
+
+func getLoggedInUsers() ([]string, error) {
```
Comment (maidul98) [inline cli/packages/cmd/user.go:151]: i think it'd be great if you can write quick function descriptions in 1-2 sentences for all functions you create. It would help future contributors to quickly understand

File: cli/packages/cmd/user.go [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+
+		err = util.WriteConfigFile(&configFile)
+		if err != nil {
+			util.HandleError(err, "")
+		}
+	},
+}
+
+var domainCmd = &cobra.Command{
+	Use:                   "domain",
```
Comment (maidul98) [inline cli/packages/cmd/user.go:83]: what do you think about putting command under `update` subcommand like this `infisical user update domain`?

File: cli/packages/util/config.go [hunk 8]
```diff
@@ -31,9 +32,25 @@ func WriteInitalConfig(userCredentials *models.UserCredentials) error {
 		return fmt.Errorf("writeInitalConfig: unable to write config file because [err=%s]", err)
 	}
 
+	//if profiles exists
+	loggedInUser := models.LoggedInUser{
+		Email:  userCredentials.Email,
+		Domain: config.INFISICAL_URL,
+	}
+	if len(existingConfigFile.LoggedInUsers) > 0 {
```
Comment (maidul98) [inline cli/packages/util/config.go:40]: i think this can be simplified to:

```
if len(existingConfigFile.LoggedInUsers) == 0 || !ConfigContainsEmail(existingConfigFile.LoggedInUsers, userCredentials.Email) {
    existingConfigFile.LoggedInUsers = append(existingConfigFile.LoggedInUsers, loggedInUser)
}
```

File: cli/packages/util/config.go [hunk 9]
```diff
@@ -31,9 +32,25 @@ func WriteInitalConfig(userCredentials *models.UserCredentials) error {
...
+		}
+	} else {
+		existingConfigFile.LoggedInUsers = append(existingConfigFile.LoggedInUsers, loggedInUser)
+	}
+
 	configFile := models.ConfigFile{
-		LoggedInUserEmail: userCredentials.Email,
-		VaultBackendType:  existingConfigFile.VaultBackendType,
+		LoggedInUserEmail:  userCredentials.Email,
+		LoggedInUserDomain: config.INFISICAL_URL,
```
Comment (maidul98) [inline cli/packages/util/config.go:51]: we set the current logged in user domain here but i don't think we update it in `LoggedInUsers` domain property  if the user email already exists. Let me know if i misunderstood

**Comment by maidul98** [comment]
Overall, excellent work @quinton11! I think the Infisical community will really love this. I left some inital feedback, tomorrow i'll do more thorough testing to make sure all all edge cases are covered and existing users won't have any issues. 

Also I think both of these methods of setting the domain should override the user set domain, what do you think?
Screenshot 2023-04-18 at 9 14 11 PM

File: cli/packages/cmd/login.go (hunk 1 above)
Comment (quinton11) [reply cli/packages/cmd/login.go:57]: Yep, my bad.

File: cli/packages/cmd/login.go (hunk 2 above)
Comment (quinton11) [reply cli/packages/cmd/login.go:286]: `idx` referring to the selected index of the options. However yes `selectedHostingOption` would be more readable. Instead of dealing with the index, comparing the selected value directly should be better. Thanks

File: cli/packages/cmd/login.go (hunk 3 above)
Comment (quinton11) [reply cli/packages/cmd/login.go:317]: Yeah, my initial thought was that will be added directly by the users. However adding a placeholder text example you suggested, as to the format they're to input would resolve the ambiguity.

File: cli/packages/cmd/login.go (hunk 4 above)
Comment (quinton11) [reply cli/packages/cmd/login.go:307]: Yeah we could provide the `default` value in the prompt object. Example: `Default: 'Example: https://my-domain-example.com`

File: cli/packages/cmd/login.go (hunk 5 above)
Comment (quinton11) [reply cli/packages/cmd/login.go:414]: Yeah this shouldn't be an issue, there is an added check for the current logged in user domain in https://github.com/Infisical/infisical/pull/481/files/be2cf54d6e138b8d82a5310be941f43310b9823f#diff-eb3a7252b09a5793a44fec4d5e033dda44f2fc86484c3d1a728bb57208058fda.

File: cli/packages/cmd/user.go (hunk 6 above)
Comment (quinton11) [reply cli/packages/cmd/user.go:151]: Yeah, thanks

File: cli/packages/cmd/user.go (hunk 7 above)
Comment (quinton11) [reply cli/packages/cmd/user.go:83]: Right, wouldn't be an issue, its more user readable

File: cli/packages/util/config.go (hunk 9 above)
Comment (quinton11) [reply cli/packages/util/config.go:51]: Right, good eye! Will amend

**Comment by quinton11** [comment]
> Overall, excellent work @quinton11! I think the Infisical community will really love this. I left some inital feedback, tomorrow i'll do more thorough testing to make sure all all edge cases are covered and existing users won't have any issues.
> 
> Also I think both of these methods of setting the domain should override the user set domain, what do you think? Screenshot 2023-04-18 at 9 14 11 PM

Thanks!
Yes I agree, however dealing with multiple profiles, the sub command was to allow users change the domain of a particular profile. What could happen is we could still use the command but check for the INFISICAL_API_URL exported variable, if there is a value, ask user to either use or override

**Comment by maidul98** [comment]
> > Overall, excellent work @quinton11! I think the Infisical community will really love this. I left some inital feedback, tomorrow i'll do more thorough testing to make sure all all edge cases are covered and existing users won't have any issues.
> > Also I think both of these methods of setting the domain should override the user set domain, what do you think? Screenshot 2023-04-18 at 9 14 11 PM
> 
> Thanks! Yes I agree, however dealing with multiple profiles, the sub command was to allow users change the domain of a particular profile. What could happen is we could still use the command but check for the INFISICAL_API_URL exported variable, if there is a value, ask user to either use or override

I'm thinking we override by default for both `INFISICAL_API_URL` exported variable and and `--domain` flag. If either of those are present, we ignore the domain for the user. This is because existing users already expect these methods of setting the domain

File: cli/packages/cmd/login.go (hunk 3 above)
Comment (maidul98) [reply cli/packages/cmd/login.go:317]: yes, good thought. Let's add the /api in the place holder text so they know

File: cli/packages/cmd/login.go (hunk 4 above)
Comment (maidul98) [reply cli/packages/cmd/login.go:307]: or maybe `https://my-self-hosted-instance.com`? Feel free to go with what you think is best for developers to understand

File: cli/packages/cmd/login.go (hunk 1 above)
Comment (maidul98) [reply cli/packages/cmd/login.go:57]: yep, no worries

**Comment by maidul98** [comment]
Shoot me a ping when you have made new changes for the comments we disccused ^ and i'll review again! Thanks again

File: cli/packages/util/config.go (hunk 8 above)
Comment (quinton11) [reply cli/packages/util/config.go]: Yep, addressed this

File: cli/packages/cmd/user.go [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+
+		//prompt user
+		profile, err := LoggedInUsersPrompt(loggedInProfiles)
+		if err != nil {
+			util.HandleError(err, "[infisical user update domain]: Prompt error")
+		}
+
+		domain := ""
+		domainQuery := true
+		if config.INFISICAL_URL_MANUAL_OVERRIDE != util.INFISICAL_DEFAULT_API_URL {
```
Comment (maidul98) [inline cli/packages/cmd/user.go:115]: Can you explain why you choose to use `INFISICAL_URL_MANUAL_OVERRIDE` variable instead of   `config.INFISICAL_URL`? I see in `GetCurrentLoggedInUserDetails`, you set `INFISICAL_URL_MANUAL_OVERRIDE=config.INFISICAL_URL` so i'm curious if we need this or if there is a use case I didn't catch

File: cli/packages/util/credentials.go [hunk 11]
```diff
@@ -87,6 +88,13 @@ func GetCurrentLoggedInUserDetails() (LoggedInUserDetails, error) {
 			SetAuthToken(userCreds.JTWToken).
 			SetHeader("Accept", "application/json")
 
+		config.INFISICAL_URL_MANUAL_OVERRIDE = config.INFISICAL_URL
```
Comment (maidul98) [inline cli/packages/util/credentials.go:91]: Can you explain why you choose to use `INFISICAL_URL_MANUAL_OVERRIDE` variable instead of   `config.INFISICAL_URL`? I see in `GetCurrentLoggedInUserDetails`, you set `INFISICAL_URL_MANUAL_OVERRIDE=config.INFISICAL_URL` so i'm curious if we need this or if there is a use case I didn't catch

File: cli/packages/cmd/user.go (hunk 10 above)
Comment (maidul98) [inline cli/packages/cmd/user.go:115]: Can you explain why you choose to use `INFISICAL_URL_MANUAL_OVERRIDE` variable instead of   `config.INFISICAL_URL`? I see in `GetCurrentLoggedInUserDetails`, you set `INFISICAL_URL_MANUAL_OVERRIDE=config.INFISICAL_URL` so i'm curious if we need this or if there is a use case I didn't catch

Comment (quinton11) [reply cli/packages/cmd/user.go:115]: Right, so with the two manual override methods, `config.INFISICAL_URL` gets updated with the user self-hosted-instance url from either `--domain` or `INFISCAL_API_URL` env var  before it hits the run function of any infisical command, now if you notice in the `PreRun` or `Run` of most commands there is a check for a user's logged in status in `GetCurrentLoggedInUserDetails`. For multi-profiles the config file holds the current logged in user as well as the user's domain. Now in here: https://github.com/Infisical/infisical/pull/481/commits/d4206cdbd82707d5ed3e655bf5e1012c1241d619#diff-eb3a7252b09a5793a44fec4d5e033dda44f2fc86484c3d1a728bb57208058fdaR93 , there is a check for if the config data structure holds a `LoggedInUserDomain` value. If there is it overrides the `config.INFISICAL_URL`. To still support the two manual override methods we don't want that. So the `config.INFISICAL_URL_MANUAL_OVERRIDE` gets assigned the `config.INFISICAL_URL` value before its updated. This is necessary because we can use that to check if there was a domain override from those two methods in the `Run` functions of the relevant commands, in this case `login` and `user update domain`.

Comment (quinton11) [reply cli/packages/cmd/user.go:115]: Then we check if there was an override by comparing `config.INFISICAL_URL_MANUAL_OVERRIDE` with the default api

Comment (maidul98) [reply cli/packages/cmd/user.go:115]: Oh i see, makes sense. So INFISICAL_URL_MANUAL_OVERRIDE holds the value defined by the two methods and is used to detect if those two methods are used. Thanks for explaining

**Review by maidul98 — APPROVED** [review]
